### PR TITLE
feat(cdk-experimental/menu): add functionality to navigate a Menu and MenuBar with a keyboard

### PR DESCRIPTION
**Pull request by andy9775**
The aria spec for Menu and MenuBar describes a robust set of instructions for navigating the Menu
using a keyboard. This implements the full aria spec for vertical and horizontal Menus along with
support for left-to-right and right-to-left layouts. Some liberty is taken with handling keyboard
events with modifier keys (home/end and escape) in order to match the existing mat-menu
implementation for those keys.

### CODE REVIEW

File: src/cdk-experimental/menu/menu-bar.ts [hunk 1]
```diff
@@ -20,18 +35,169 @@ import {CDK_MENU, Menu} from './menu-interface';
...
+        break;
+
+      case HOME:
+      case END:
+        if (!hasModifierKey(event)) {
+          event.preventDefault();
+          event.keyCode === HOME ? this.focusFirstItem('keyboard') : this.focusLastItem('keyboard');
+        }
+        break;
+
```
Comment (jelbourn) [inline src/cdk-experimental/menu/menu-bar.ts:126]: FYI we just support for home and end in `ListKeyManager` this past week

**Comment by jelbourn** [comment]
@crisbeto I'd love to get your input on this approach; I've had my head in the problem for too long

File: src/cdk-experimental/menu/menu-bar.ts [hunk 2]
```diff
@@ -20,18 +35,169 @@ import {CDK_MENU, Menu} from './menu-interface';
   selector: '[cdkMenuBar]',
   exportAs: 'cdkMenuBar',
   host: {
+    '(keydown)': '_handleKeyEvent($event)',
+    '(focus)': 'focusFirstItem()',
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-bar.ts:39]: If we're redirecting focus anyway, couldn't we remove the `tabindex` from this this and let the user tab into the first item directly?

File: src/cdk-experimental/menu/menu-bar.ts [hunk 3]
```diff
@@ -20,18 +35,169 @@ import {CDK_MENU, Menu} from './menu-interface';
...
+        // For a horizontal menu if the left/right keys were clicked, or a vertical menu if the
+        // up/down keys were clicked: if the current menu is open, close it then focus and open the
+        // next  menu.
+        if (
+          (this._isHorizontal() && horizontalArrows) ||
+          (!this._isHorizontal() && !horizontalArrows)
+        ) {
+          event.preventDefault();
+
+          const prevIsOpen = this._keyManager.activeItem?.isMenuOpen();
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-bar.ts:108]: It looks like you're repeating `this._keyManager.activeItem` in almost all of the switch cases. Maybe move it out into a variable at the top?

File: src/cdk-experimental/menu/menu-bar.ts [hunk 4]
```diff
@@ -20,18 +35,169 @@ import {CDK_MENU, Menu} from './menu-interface';
...
+      this._keyManager.withHorizontalOrientation(this._dir.value);
+    } else {
+      this._keyManager.withVerticalOrientation();
+    }
+  }
+
+  /** Subscribe to the MenuStack close and empty observables. */
+  private _subscribeToMenuStack() {
+    this._menuStack.close().subscribe((item: MenuStackItem) => this._closeOpenMenu(item));
+    this._menuStack.empty().subscribe((event: FocusNext) => this._toggleOpenMenu(event));
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-bar.ts:155]: It doesn't look like these are unsubscribing.

File: src/cdk-experimental/menu/menu-bar.ts [hunk 5]
```diff
@@ -20,18 +35,169 @@ import {CDK_MENU, Menu} from './menu-interface';
...
+
+  /**
+   * Close the open menu if the current active item opened the requested MenuStackItem.
+   * @param item the MenuStackItem requested to be closed.
+   */
+  private _closeOpenMenu(item: MenuStackItem) {
+    if (item === this._keyManager.activeItem?.getMenu()) {
+      this._keyManager.activeItem.getMenuTrigger()?.closeMenu();
+      this._keyManager.setFocusOrigin('keyboard');
+      this._keyManager.setActiveItem(this._keyManager.activeItem);
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-bar.ts:166]: Repeating `this._keyManager` won't minify with anything aside from Closure compiler. If you have it multiple times within the same function, it's better to save it in a variable.

File: src/cdk-experimental/menu/menu-item-trigger.ts [hunk 6]
```diff
@@ -72,34 +88,96 @@ export class CdkMenuItemTrigger implements OnDestroy {
...
+      this.isMenuOpen() ? this.closeMenu() : this.openMenu();
+    }
+  }
+
+  /** Open the attached menu. */
+  openMenu() {
+    if (!this.isMenuOpen()) {
+      this.opened.next();
+
+      this._overlayRef = this._overlay.create(this._getOverlayConfig());
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-item-trigger.ts:100]: This will recreate the overlay every time you open it. You can reuse it like this:

```
if (!this._overlayRef) {
  this._overlayRef = this._overlay.create(this._getOverlayConfig());
}

this._overlayRef.attach(this._getPortal());
```

File: src/cdk-experimental/menu/menu-item-trigger.ts [hunk 7]
```diff
@@ -72,34 +88,96 @@ export class CdkMenuItemTrigger implements OnDestroy {
...
+      case ENTER:
+        event.preventDefault();
+        this.toggle();
+        this.menuPanel?._menu?.focusFirstItem('keyboard');
+        break;
+
+      case RIGHT_ARROW:
+        if (this._isParentVertical()) {
+          event.preventDefault();
+          if (this._directionality.value === 'ltr') {
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-item-trigger.ts:150]: The `_directionality` should be decorated with `@Optional`, because we aren't guaranteed for it to exist. Also generally we format direction checks like this, because LTR is the most common case and it handles undefined values better:

```
if (direction === 'rtl') {
  // do RTL stuff
} else {
  // do LTR stuff
}
```

File: src/cdk-experimental/menu/menu-item.ts [hunk 8]
```diff
@@ -35,21 +50,112 @@ export class CdkMenuItem {
...
+
+  /** Get the MenuItemTrigger associated with this element. */
+  getMenuTrigger(): CdkMenuItemTrigger | undefined {
+    return this._menuTrigger;
+  }
+
+  /** Get the label for this element which is required by the FocusableOption interface. */
+  getLabel(): string {
+    // TODO(andy): implement a more robust algorithm for determining nested text
+    return this._elementRef.nativeElement.innerText;
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-item.ts:109]: `innerText` will trigger a reflow, because it has some logic where it excludes hidden text. We should use `textContent` here.

File: src/cdk-experimental/menu/menu-stack.ts [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+/**
+ * MenuStack allows subscribers to listen for close events (when a MenuStackItem is popped off
+ * of the stack) in order to perform closing actions. Upon the MenuStack being empty it emits
+ * from the `empty` observable specifying the next focus action which the listener should perform
+ * as requested by the closer.
+ */
+export class MenuStack {
+  /** All MenuStackItems tracked by this MenuStack. */
+  private readonly _elements: MenuStackItem[] = [];
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-stack.ts:30]: I think that using a stack/array that pretends to be a stack here will run into the same issue as when we tried to keep track of menu items ourselves in `mat-menu`. If a menu pushes itself onto the stack, gets destroyed by an `ngIf` and then gets re-added, the stack will be out of order. E.g. if we have `[1, 2, 3]`, we remove 2 and then we re-add it, the stack will have `[1, 3, 2]`, whereas the DOM will still be `[1, 2, 3]`.

That being said, I'm having a hard time visualizing when the registration happens. If it's only while a stack of menus is open, we'll probably be fine.

File: src/cdk-experimental/menu/menu-stack.ts [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+  /**
+   * Pop off all MenuStackItems and emit each one on the `close` observable one by one.
+   * @param focusNext the event to emit on the `empty` observable once the stack is emptied. Does
+   * not emit if the stack was initially empty.
+   */
+  closeAll(focusNext?: FocusNext) {
+    if (this._elements.length) {
+      while (this._elements.length) {
+        const menuStackItem = this._elements.pop();
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-stack.ts:80]: It looks like you're repeating a bit of the logic from `closeLatest`. Could it be reduced to this?

```
while (this._elements.length) {
  this.closeLatest();
}
```

File: src/cdk-experimental/menu/menu-stack.ts [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+    }
+  }
+
+  /**
+   * Pop off all MenuStackItems and emit each one on the `close` observable one by one.
+   * @param focusNext the event to emit on the `empty` observable once the stack is emptied. Does
+   * not emit if the stack was initially empty.
+   */
+  closeAll(focusNext?: FocusNext) {
+    if (this._elements.length) {
```
Comment (crisbeto) [inline src/cdk-experimental/menu/menu-stack.ts:78]: This check isn't necessary since the loop won't start if the array is empty.

Comment (andy9775) [reply src/cdk-experimental/menu/menu-stack.ts:78]: yeah, I don't want to emit on the `_empty` subject if we call closeAll with an empty stack. Seemed cleaner than having a `cont shouldEmit = this._elements.length > 0` variable and checking it later on before calling `next`

File: src/cdk-experimental/menu/menu-item.spec.ts [hunk 12]
```diff
@@ -3,60 +3,94 @@ import {ComponentFixture, TestBed, async} from '@angular/core/testing';
...
-  }));
-
-  beforeEach(() => {
-    fixture = TestBed.createComponent(SingleMenuItem);
-    fixture.detectChanges();
-
-    button = fixture.debugElement.query(By.directive(CdkMenuItem)).injector.get(CdkMenuItem);
-    nativeButton = fixture.debugElement.query(By.directive(CdkMenuItem)).nativeElement;
-  });
+  describe('without enclosing elements', () => {
```
Comment (teflonwaffles) [inline src/cdk-experimental/menu/menu-item.spec.ts:10]: I thought this meant, without elements *around* the menu item. Could you reword this to be clearer (eg. 'with no complex inner elements', to match the other suite).

File: src/cdk-experimental/menu/menu-bar.ts (hunk 2 above)
Comment (andy9775) [reply src/cdk-experimental/menu/menu-bar.ts:39]: yeah, I didn't want to pollute this PR with too much tab focus logic. I'll have a follow on which sets the tabindex of the first menu item in the menu bar to 0 and handles `changes` as well.

File: src/cdk-experimental/menu/menu-stack.ts [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+  private readonly _elements: MenuStackItem[] = [];
+
+  /** Emits the element which was popped off of the stack when requested by a closer. */
+  private readonly _close: Subject<MenuStackItem> = new Subject();
+
+  /** Emits once the MenuStack has become empty after popping off elements. */
+  private readonly _empty: Subject<FocusNext> = new Subject();
+
+  /** Observable which emits the MenuStackItem which has been requested to close. */
+  close(): Observable<MenuStackItem> {
```
Comment (teflonwaffles) [inline src/cdk-experimental/menu/menu-stack.ts:39]: We have an internal guideline to use typing to restrict the type when exposing a subject as an observable, instead of explicitly calling asObservable(). @jelbourn , is this also true in the CDK codebase, or do you prefer to call asObservable()?

File: src/cdk-experimental/menu/menu-stack.ts (hunk 10 above)
Comment (andy9775) [reply src/cdk-experimental/menu/menu-stack.ts:80]: Long story short, no. It has to do with the Subject which results in loop

File: src/cdk-experimental/menu/menu-stack.ts (hunk 11 above)
Comment (andy9775) [reply src/cdk-experimental/menu/menu-stack.ts:78]: yeah, but I don't want to emit if the stack was initially empty. Of course I could do something like 
```js
const shouldEmit = this._elements.length > 0;
while(...) {...}
if (shouldEmit) {this._empty.next(focusNext)}
```

But this seemed cleaner IMO

File: src/cdk-experimental/menu/menu-stack.ts (hunk 9 above)
Comment (andy9775) [reply src/cdk-experimental/menu/menu-stack.ts:30]: Hmm. Not sure if I follow. If we remove 2 that means we closed 2 and closing 2 should close all submenus. Hence `[1,2,3]` would result in `[1]` after removing 2.

**Comment by andy9775** [comment]
@teflonwaffles @jelbourn @crisbeto Feedback should be addressed.

File: src/cdk-experimental/menu/menu-stack.ts (hunk 9 above)
Comment (jelbourn) [reply src/cdk-experimental/menu/menu-stack.ts:34]: Yeah, I _think_ that if you ensure any menu closing also closes all of its sub-menus, this should be okay. (Unless there's something a user might do that I'm not imagining)

File: src/cdk-experimental/menu/menu-stack.ts (hunk 13 above)
Comment (jelbourn) [reply src/cdk-experimental/menu/menu-stack.ts]: Yeah, we could do that here; that is probably better

File: src/cdk-experimental/menu/menu-stack.ts (hunk 9 above)
Comment (andy9775) [reply src/cdk-experimental/menu/menu-stack.ts:34]: Was already merged in before I had a chance to address this. I'll address with a follow on pr.

File: src/cdk-experimental/menu/menu-stack.ts (hunk 13 above)
Comment (andy9775) [reply src/cdk-experimental/menu/menu-stack.ts]: Was already merged in before I had a chance to address, will include fix in a follow on pr.

File: src/cdk-experimental/menu/menu-bar.ts (hunk 2 above)
Comment (andy9775) [reply src/cdk-experimental/menu/menu-bar.ts:39]: Considering this further, the current implementation will set focus on the first non-disabled item and keeps focus on the MenuBar if all elements are disabled which is the expected behaviour. Any changes would essentially do the same. The only difference being that the tab index of the first menu item would be 0 rather than -1 as is here. Thoughts?
